common/dal-context/src/context.rs

```diff
@@ -110,26 +111,38 @@ impl DalContext {
 
 impl Layer for DalContext {
     fn layer(&self, inner: Arc<dyn Accessor>) -> Arc<dyn Accessor> {
-        Arc::new(DalContext::new(inner))
+        Arc::new(DalContext {
+            inner: Some(inner),
+            metrics: self.metrics.clone(),
+        })
     }
 }
 
 #[async_trait]
 impl Accessor for DalContext {
     async fn read(&self, args: &OpRead) -> DalResult<Reader> {
-        // TODO(xuanwo): Implement context callback reader to collect metrics.
-        self.inner.read(args).await
+        let metrics = self.metrics.clone();
+
+        // TODO(xuanwo): Maybe it's better to move into metrics.
+        self.inner.as_ref().unwrap().read(args).await.map(|reader| {
```

Comment:
is it possible to eliminate this `unwrap` (and other `unwrap` in non-unit-testing codes)?

Reply:
There is no other way to access the `inner` outside `DalContext`, I believe it's safe here. But I do understand your concern, I will take it into consideration. How about resolving it in the next PRs?

Reply:
Tracked in https://github.com/datafuselabs/databend/issues/3677

Reply:
> But I do understand your concern, I will take it into consideration.  

`unwrap` may panic. returns an Error might be better (at the Dal Layer)...

> How about resolving it in the next PRs?

Sure